---
title: "Cluster TNBC Data"
params:
  geo.dataset: GSE76250
  top.gene.count: 500
  old.microarray.platform.name: 'Affymetrix HG-U133A'
  old.microarray.platform.id: 'affy_hg_u133a'
  qgs.length: 50
output:
  html_document: default
---

The purpose of this RMarkdown document is to create Query Gene Signatures that represent the diseased
state of Triple-Negative Breast Cancer.


# Downloading the data
We're going to use a dataset of TNBC data from another article that was clustering TNBC data from The GEO dataset `r params$geo.dataset`.

For this we require Bioconductor's `GEOquery` to download the files:

```{r}
library('GEOquery')
dir.create('geodata')
tnbc.gene.set <- getGEO(params$geo.dataset, destdir="geodata")[[1]]
```

```{r}
case.row.nums <- which(pData(tnbc.gene.set)$source_name_ch1 == 'breast_TNBC')
case.ids <- pData(tnbc.gene.set)$geo_accession[case.row.nums]
control.row.nums <- which(pData(tnbc.gene.set)$source_name_ch1 == 'breast_normal tissue')
control.ids <- pData(tnbc.gene.set)$geo_accession[control.row.nums]
```

The resulting ExpressionSet contains `r ncol(tnbc.gene.set@assayData$exprs)` samples, of which there are 
`r length(case.ids)` TNBC expression profiles along with `r length(control.ids)` expression profiles for non-cancerous
adjacent tissue. The case and control samples are needed for performing our differential expression analysis to deduce
which genes we should use to perform the clustering. When performing the clustering, all control samples will be 
removed.

Before we perform the analysis, there are a number of steps required to prepare the data.

Most significantly, as this is a GEO GSE as opposed to a GDS, the microarray data has not already been normalized.
For this we can apply quantile normalization using the `limma` package:

```{r}
require('limma')

exprs(tnbc.gene.set) <- normalizeBetweenArrays(exprs(tnbc.gene.set))
```

Furthermore, as these microarray data were measured using the Affymetrix GeneChip&reg; Human Transcriptome Array 2.0,
there are duplicate probe IDs on the microarray which map to the same genes.

In accordance with the original Liu et al paper <!-- add reference here -->, we will remove genes with what is deemed a nominal small standard deviation, namely $\sigma \leq .65$ or $\sigma^2 \leq .65$

```{r}
require('genefilter')
# also download the annotation package for this microarray platform
require('hta20transcriptcluster.db')

# assign correct annotation package name to this GSE
annotation(tnbc.gene.set) <- 'hta20transcriptcluster.db'

min.variance <- .65 ^ 2

#perform filtering to retain probes with highest IQR
tnbc.gene.set <- genefilter::nsFilter(tnbc.gene.set, remove.dupEntrez =T,
                                           var.filter=T, var.cutoff=min.variance,
                                          filterByQuantile=F)

removed.element.details <- tnbc.gene.set[[2]]

tnbc.gene.set <- tnbc.gene.set[[1]]
```

Following the filtering, `r removed.element.details$numDupsRemoved` duplicate genes were removed,
`r removed.element.details$numLowVar` genes were removed due to low variance and `r removed.element.details$numRemoved.ENTREZID` were removed for having no entrez id <!-- perhaps stop this happening by 
setting require.entrez=F!!!!`-->.

# Differential Expression Analysis

We will use limma to perform differential expression analysis between the TNBC expression profiles and the 
adjacent tissue expression profiles. Firstly, let's calculate the vector of `outcome measurements` for each sample 
required by limma:

```{r}
col.names <- colnames(exprs(tnbc.gene.set))
label.fn <- function(sample.name) {
  if (sample.name %in% case.ids) {
    return(1)
  } else {
    return(2)
  }
}
outcome <- sapply(col.names, FUN=label.fn)
```

With these data we can perform the differential expression analysis. We will perform the differential expression analysis using the limma R package:

```{r}
design <- model.matrix(~ 0+factor(outcome))
colnames(design) <- c('case', 'control')
fit <- lmFit(tnbc.gene.set, design)
fit <- eBayes(fit)
top.genes <- topTable(fit, coef=2, number=params$top.gene.count, adjust.method="fdr", p.value=0.05,lfc=0)
# remove '.1' from end of probset_id column
top.genes$probeset_id <- substr(top.genes$probeset_id, 1, nchar(as.character(top.genes$probeset_id)) - 2)
```

# Creating a global TNBC Query Gene Signature

In order for us to create a Query Gene Signature that's compatible with the the reference profiles, we need 
to convert the Affymetrix GeneChip&reg; Human Transcriptome Array 2.0 probe IDs into the equivalent probe IDs
from the `r params$old.microarray.platform.name` microarray platform. This is a necessary step as the referenece profiles describing 
drug action used the latter platform.

To do this, we use the `biomaRt` package:

```{r biomaRt map ids}
library('biomaRt')

hta.ids <- as.character(top.genes$probeset_id)

mart <- useDataset('hsapiens_gene_ensembl', useMart('ensembl'))
probe.id.mapping <- getBM(filters='affy_hta_2_0', attributes= 
                   c("ensembl_gene_id", params$old.microarray.platform.id, 'affy_hta_2_0', 'description'),
                 values=hta.ids,mart= mart)
```

Having discovered all probe IDs that map to the `r params$old.microarray.platform.name` platform, we can join these with 
or table of top genes:

```{r merge dataframe with u133a probe ids}
top.genes.with.probeids <- merge(probe.id.mapping, top.genes, by.x = 'affy_hta_2_0', by.y = 'probeset_id')
```

To follow, we can filter out the rows where none of there are no probe IDs on the `r params$old.microarray.platform.name` 
platform.

```{r filter out empty u133a}
empty.u133a.id.rows <- which(is.na(top.genes.with.probeids$affy_hg_u133a) | top.genes.with.probeids$affy_hg_u133a == '')
top.genes.with.probeids <- top.genes.with.probeids[-empty.u133a.id.rows,]
```

Furthermore in the mapping there will be duplicate probe IDs due to multiple mappings found in the Ensembl database. We should also remove these before creating our query gene signature.

```{r remove duplicate probe ids}
non.duplicate.rows <- which(!duplicated(top.genes.with.probeids$affy_hg_u133a))
top.genes.with.probeids <- top.genes.with.probeids[non.duplicate.rows,]
```

## Creating the global Query Gene Signature (Unordered and Ordered)
We will take the $$N = `r params$qgs.length`$$ most perturbed genes (highest log fold change) to create our query gene signatures.

```{r top 50 perturbed genes}
top.genes.ordered.by.logFC <- top.genes.with.probeids[order(abs(top.genes.with.probeids$logFC), decreasing = T),]
# only to N
top.n.ordered.by.logFC <- top.genes.ordered.by.logFC[1:params$qgs.length,]
```

### Unordered QGS
Next we will create the unordered query gene signature by taking the signs of each of the `logFC` and making them `-1` or `+1`. 

```{r unordered global qgs}
unorderd.global.qgs <- sign(top.n.ordered.by.logFC$logFC)
names(unorderd.global.qgs) <- top.n.ordered.by.logFC$affy_hg_u133a
unorderd.global.qgs
```
### Ordered QGS
To create the orderd query gene signature we multiply the rank of each logFC by the unordered query gene signature. As 
`top.n.ordered.by.logFC` is already ordered, we can just multiple each of the probes by the values from 1 to 
`r params$qgs.length`.

```{r ordered global qgs}
ordered.global.qgs <- unorderd.global.qgs * seq(1,params$qgs.length)
ordered.global.qgs
```

<!-- TODO - compare to RNAseq data qgs in report -->


# K-means Clustering by significantly differentially expressed genes

Albeit we have successfully created a Query Gene Signature that globally describes the state of TNBC, it might be
that we get better results by identifying subtypes of TNBC by using clustering algorithms.

While the significantly differentially expressed probes were unable to be used to synthesise a query gene signature for TNBC, we can use these same probe IDs as the basis for clustering our data. K-means clustering will be performed
to find subtypes of triple-negative breast cancer.

Firstly, let's take the names of the significantly differentially expressed probe ids create an expression matrix 
with only these probe IDs. This will serve as input to our clustering algorithm.

```{r create subset expression matrix}
# limma version
diff.expressed.probe.ids <- as.character(top.genes$probeset_id)

# need to add in '.1' to end to compare with expression matrix
probe.ids.with.1 <- sapply(diff.expressed.probe.ids, function(probe.id) return(paste0(probe.id,'.1')))

rows <- which(row.names(exprs(tnbc.gene.set)) %in% probe.ids.with.1)

# we only want non control id expression values
case.ids <- as.character(case.ids)
diff.expressed.expr.set <- exprs(tnbc.gene.set)[rows,case.ids]

# need to transpose for kmeans - rows = datapoints
diff.expr.transposed <- t(diff.expressed.expr.set)
```

Now we have the expression matrix for all samples that are differentially expressed we can go about clustering. But 
first we need to find the ideal number of clusters, `K*`. This can be approximated by using Rob Tibshirani's *gap statistic*:

<!-- TODO: try using kmeansW from https://cran.r-project.org/web/packages/FactoClass/FactoClass.pdf allows weighting of differential variables -->

```{r}
gap.statistic <- cluster::clusGap(diff.expr.transposed, cluster::pam,K.max=10, B=500)
plot(gap.statistic)
print(gap.statistic, method='Tibs2001SEmax')
```

We can see, according to the Tibshirani method, the optimal number of clusters (and therefore the ideal number of 
Triple-Negative Breast Cancer subtypes) from this sample of data is 4. This reflects the results found by Liu et al in their 
paper using the same datapoints albeit they performed a different clustering method called Consensus Clustering.

```{r perform-clustering}
clusters <- cluster::pam(diff.expr.transposed, k=4)
```

The data segregate into four clusters with the following distribution:

```{r distribution-of-clusters}
table(clusters$clustering)
```

Now that the expression values have been clustered into the optimal number of clusters, $K*$, we can now create query gene 
signatures for each subtype identified which will later be used to perform Gene Expression Connectivity Map such that we mightf find candidate therapies for each breast cancer subtype.

<!-- # Consensus Clustering
 
TODO use R package ConsensusClustPlus to find optimal numeber of clusters - method similar to original paper 
(Liu et al, 2016)

https://www.ncbi.nlm.nih.gov/pubmed/20427518
As no clusters were identified using a k-means clustering approach with Tibshirani's gap statistic, another approach is to 
use a consensus clustering approach implemented by the Bioconductor package ConsensusClusterPlus.

Consensus works by creating random subsamples of the original dataset and to perform a particular clustering algorithm (e.g. PAM or K-means) over multiple values for K. A consensus value is calculated for each value of K which indicates how cleanly
all of the clusters and the optimal K is evaluated by plotting an empirical cumulative distribution function.

## Perform Consensus Clustering with ConsensusClusterPlus
```{r}
!dir.exists('consensus_clust') &&  dir.create('consensus_clust')
library(ConsensusClusterPlus)
results = ConsensusClusterPlus(diff.expr.transposed,maxK=6,reps=50,pItem=0.8,pFeature=1,
title='consensus_clust',clusterAlg="hc",distance="pearson",seed=1262118388.71279,plot="png")
```
-->

<!-- TOOD add section discussing pathways involved in these subtypes -->
# Create QGSs for Subtypes

Now we have identifed these subtypes, we can partition our data according to the subtypes the samples segregate into.
From this we can create individual Query Gene Signatures for each subtype. Such that we can perform differential 
expression analysis on each of these subsets, it befits us to reintroduce the non TNBC samples to each of these 
expression matrices.

```{r partition data for subtypes}
tnbc.expr.data <- exprs(tnbc.gene.set)

create.clust.subset <- function(clust.num) {
  clust.sample.ids <- names(which(clusters$clustering == clust.num))
  
  sample.indices <- which(colnames(tnbc.expr.data) %in% clust.sample.ids)
  control.indices <- which(design[,2] == 1)
  # put the controls at the end
  return(tnbc.expr.data[,c(sample.indices, control.indices)])
}

clust.subsets <- lapply(X = list(clust1=1,clust2=2,clust3=3,clust4=4), FUN = create.clust.subset)
```

Now that we have created expression matrices for each of the four subsets we can now perform differential expression 
analysis on each as we did before globally. We will reuse the design matrix as bef

```{r subtype differential expression analysis}
find.top.genes <- function(expr.mat) {
  num.control <- length(which(design[,2] == 1))
  num.case <- ncol(expr.mat) - num.control
  
  control.flags <- c(rep(0, num.case), rep(1, num.control))
  case.flags <- c(rep(1, num.case), rep(0, num.control))

  clust.design.mat <- cbind(case.flags, control.flags)
  names(clust.design.mat) <- c('case', 'control')
  fit <- lmFit(expr.mat, clust.design.mat)
  fit <- eBayes(fit)
  return(topTable(fit, coef=2, number=params$top.gene.count, adjust.method="fdr", p.value=0.05,lfc=0))
}

clust.top.genes <- lapply(X=clust.subsets, FUN = find.top.genes)

# remove '.1' from end of probset_id column
clust.top.genes.mapped.probes <- lapply(X=clust.top.genes, FUN= function(clust.df) {
  new.rownames <- substr(rownames(clust.df), 1, nchar(rownames(clust.df)) - 2)
  rownames(clust.df) <- new.rownames
  return(clust.df)
})
```

We can observe the overlap of the top `r params$top.gene.count` differentially expressed genes among the different clusters to see how 
similar or dissimilar the clusters are.

```{r subtype venn diagram sd genes}
library(VennDiagram)
clust.sd.genes <- lapply(X = clust.top.genes.mapped.probes, FUN = function(exp.mat) rownames(exp.mat))
venn.diagram(clust.sd.genes, "ClusterSDGeneVennDiagram.tiff")
```

![Venn Diagram](./ClusterSDGeneVennDiagram.tiff)

## Map cluster Affymetrix HTA 2 to `r params$old.microarray.platform.name` for each cluster

<!-- TODO add some text corresponding to creation of global qgs -->

```{r map probes for each cluster}

map.probe.ids <- function(top.gene.df) {
  gene.ids <- row.names(top.gene.df)
  mart <- useDataset('hsapiens_gene_ensembl', useMart('ensembl'))
  summary.clust <- getBM(filters='affy_hta_2_0', attributes= 
               c("ensembl_gene_id", params$old.microarray.platform.id, 'affy_hta_2_0', 'description'),
               values=gene.ids,mart= mart)

  return(summary.clust)
}

mapped.probe.ids.top.genes <- lapply(X=clust.top.genes.mapped.probes, FUN = map.probe.ids)

```


```{r merge cluster dataframes with u133a probe ids}

merge.probes <- function(probe.id.mappings, clust.top.genes) {
  return(merge(probe.id.mapping, clust.top.genes, by.x = 'affy_hta_2_0', by.y = 'row.names'))
}

clust.top.genes.with.mapped.probe.ids <- lapply(FUN = function(idx) {
  mappings <- mapped.probe.ids.top.genes[[idx]] 
  top.g    <- clust.top.genes.mapped.probes[[idx]]
  return(merge.probes(mappings, top.g))
}, X = 1:length(clust.top.genes.mapped.probes))

names(clust.top.genes.with.mapped.probe.ids) <- names(clust.top.genes.mapped.probes)
```

```{r remove duplicate and empty for clusters}

remove.empty.and.duplicate.u133a <- function(top.genes.df) {
  remove.empty.u133a <- function(top.genes.df) {
    empty.u133a.rows <- which(is.na(top.genes.df$affy_hg_u133a) | top.genes.df$affy_hg_u133a == '')
    return(top.genes.df[-empty.u133a.rows,])
  }
  
  remove.duplicate.u133a <- function(top.genes.df) {
    non.duplicate.rows <- which(!duplicated(top.genes.df$affy_hg_u133a))
    return(top.genes.df[non.duplicate.rows,])
  }
  
  return(remove.duplicate.u133a(remove.empty.u133a(top.genes.df)))
}

clust.top.genes.with.mapped.probe.ids.no.duplicate <-
  lapply(FUN = remove.empty.and.duplicate.u133a, X = clust.top.genes.with.mapped.probe.ids)
```

## Creating Query Gene Signatures for each cluster (Unordered and Ordered)
Again, we will take the $$N = `r params$qgs.length`$$ most perturbed genes (highest log fold change) to create our query gene signatures.

```{r top 50 perturbed genes for clusters}
top.N.logFC.genes.ordered <- function(mapped.probe.ids) {
  ordered.by.logFC <- mapped.probe.ids[order(abs(mapped.probe.ids$logFC), decreasing = T),]
  return(ordered.by.logFC[1:params$qgs.length,])
}

top.N.ordered.clusts <- lapply(FUN = top.N.logFC.genes.ordered, X = clust.top.genes.with.mapped.probe.ids.no.duplicate)

```

### Unordered QGSs
As before with the global query gene signature, we will create the unordered query gene signature by taking the signs of each of the `logFC` and making them `-1` or `+1`. 

```{r unordered qgs for cluster}
create.unordered.qgs <- function(top.n.ordered.genes) {
  unordered.qgs <- sign(top.n.ordered.genes$logFC)
  names(unordered.qgs) <- top.n.ordered.genes$affy_hg_u133a
  return(unordered.qgs)
}

cluster.unordered.qgs <- lapply(FUN = create.unordered.qgs, X = top.N.ordered.clusts)
cluster.unordered.qgs
```

### Ordered QGSs

```{r ordered qgs for cluster}
create.ordered.qgs <- function(unordered.qgs) {
  ordered <- seq(1,params$qgs.length)
  names(ordered) <- names(unordered.qgs)
  ordered
}

cluster.ordered.qgs <- lapply(FUN = create.ordered.qgs, X = cluster.unordered.qgs)
cluster.ordered.qgs
```


<!-- TODO 
* COMPLETE merge top genes with mapped probe ids
* DONE filter out empty / duplicate
* create qgs ordered/unordered for each cluster
* write another script to run against QUADrATiC
* -->
